### PR TITLE
Quote v2: add migration

### PR DESCRIPTION
**Pull request by oandregal**
Part of https://github.com/WordPress/gutenberg/issues/15486
Related https://github.com/WordPress/gutenberg/pull/25892

## What?

This PR adds the migration from quote v1 to quote v2.

## Why?

We need to support existing content.

## How?

- Migrates v1 versions of the block by declaring deprecations.
- Migrates blocks created from templates provided to CPT by means of a hook in the edit function.

## Testing Instructions

Verify that the deprecation works:

- Disable the v2 quote in "Gutenberg > Experiments".
- Create some post with quote content. Here's an example that covers all cases:

```html
<!-- wp:quote -->
<blockquote class="wp-block-quote"><!-- wp:paragraph -->
<p>Single p and cite</p>
<!-- /wp:paragraph --><cite>Cite</cite></blockquote>
<!-- /wp:quote -->

<!-- wp:quote -->
<blockquote class="wp-block-quote"><!-- wp:paragraph -->
<p>multiple</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>paragraph</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>and</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>cite</p>
<!-- /wp:paragraph --><cite>cite</cite></blockquote>
<!-- /wp:quote -->

<!-- wp:quote -->
<blockquote class="wp-block-quote"><!-- wp:paragraph -->
<p>multiple</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>paragraph</p>
<!-- /wp:paragraph --></blockquote>
<!-- /wp:quote -->

<!-- wp:quote -->
<blockquote class="wp-block-quote"><!-- wp:paragraph -->
<p></p>
<!-- /wp:paragraph --><cite>cite</cite></blockquote>
<!-- /wp:quote -->

<!-- wp:quote -->
<blockquote class="wp-block-quote"><!-- wp:paragraph -->
<p></p>
<!-- /wp:paragraph --></blockquote>
<!-- /wp:quote -->
```

- Enable back the v2 quote and load the post.
- The expected result is that all quotes are properly migrated to the v2 forrmat.

https://user-images.githubusercontent.com/583546/160582647-f876f1e0-c75f-40fc-af4c-0f0b322eb967.mp4

Verify that templates provided to CPT are migrated:

- Register a template for a CPT. For example, by adding the following code to `gutenberg.php`:

```php
function myplugin_register_template() {
    $post_type_object = get_post_type_object( 'post' );
    $post_type_object->template = array(
        array( 'core/quote', array(
			'value'    => '<p>First</p><p>Second</p>',
			'citation' => 'Howdy',
		) ),
    );
}
add_action( 'init', 'myplugin_register_template' );
```

- Make sure the v2 quote is enabled.
- Create a new post and verify that the quote template is properly converted to the v2.

https://user-images.githubusercontent.com/583546/160582608-fea0b611-de78-43b5-bfa0-9cd4fc6669be.mp4

### CODE REVIEW

File: packages/block-library/src/quote/v2/index.js [hunk 1]
```diff
@@ -30,37 +30,7 @@ const settings = {
...
-
-/**
- * This function updates the attributes for the quote v2.
- * This should be moved to block.json when v2 becomes the default.
- *
- * @param {Array}  blockSettings The settings of the block to be registered.
- * @param {string} blockName     The name of the block to be registered.
- * @return {Array} New settings.
- */
-function registerQuoteV2Attributes( blockSettings, blockName ) {
```
Comment (oandregal) [inline packages/block-library/src/quote/v2/index.js:45]: Had to remove the filter because it was affecting the deprecation, preventing it from having the `value` attribute to migrate. The result of this change is that when v2 is enabled, the `value` attribute will still be present.

We need to remove the attribute from the `block.json` when v2 becomes the default ([see](https://github.com/WordPress/gutenberg/pull/25892#issuecomment-1076686005)).

File: packages/block-library/src/quote/v2/edit.js [hunk 2]
```diff
@@ -9,24 +9,66 @@ import {
...
-	attributes: { citation },
+	attributes,
 	setAttributes,
 	isSelected,
 	insertBlocksAfter,
 	clientId,
 } ) {
+	const { citation } = attributes;
+
+	useMigrateOnLoad( attributes, clientId );
```
Comment (oandregal) [inline packages/block-library/src/quote/v2/edit.js:72]: This is using the same technique than the list v2 uses https://github.com/WordPress/gutenberg/pull/39799/files#diff-8ca208404bc35d0549ee68bfdb5c87a0d1fa101ddcfad1095d9a3c0e171b0272

File: packages/block-library/src/quote/v2/index.js (hunk 1 above)
Comment (youknowriad) [reply packages/block-library/src/quote/v2/index.js:45]: > We need to remove the attribute from the block.json when v2 becomes the default

I don't think we can remove it until the `useMigrateOnLoad` is removed.

File: packages/block-library/src/quote/v2/edit.js [hunk 3]
```diff
@@ -9,24 +9,66 @@ import {
...
+		// As soon as the block is loaded, migrate it to the new version.
+
+		if ( ! attributes.value ) {
+			// No need to migrate if it doesn't have the value attribute.
+			return;
+		}
+
+		const [ newAttributes, newInnerBlocks ] = migrateToQuoteV2(
+			attributes
+		);
```
Comment (youknowriad) [inline packages/block-library/src/quote/v2/edit.js:48]: Any thoughts on adding a `deprecated` call here. I considered adding one for the list block too.

File: packages/block-library/src/quote/v2/deprecated.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+export const migrateToQuoteV2 = ( attributes ) => {
+	const { value } = attributes;
+
+	return [
+		{
+			...omit( attributes, [ 'value' ] ),
+		},
+		value
+			? parseWithAttributeSchema( value, {
```
Comment (youknowriad) [inline packages/block-library/src/quote/v2/deprecated.js:26]: Interesting, I guess I could have used the same technique in the list block. Seems simpler. Is this what you had in mind @jorgefilipecosta

File: packages/block-library/src/quote/v2/deprecated.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+ */
+import { createBlock, parseWithAttributeSchema } from '@wordpress/blocks';
+import { RichText, useBlockProps } from '@wordpress/block-editor';
+
+/**
+ * Internal dependencies
+ */
+import deprecationsForV1Block from '../deprecated';
+
+export const migrateToQuoteV2 = ( attributes ) => {
```
Comment (youknowriad) [inline packages/block-library/src/quote/v2/deprecated.js:18]: Do you think it's worth unit testing this function like I did in the list block?

Comment (oandregal) [reply packages/block-library/src/quote/v2/deprecated.js:18]: Makes sense https://github.com/WordPress/gutenberg/pull/39844/commits/2bdddf11d77c5cf18a31183f003240c5a489d20d

File: packages/block-library/src/quote/v2/edit.js (hunk 3 above)
Comment (oandregal) [reply packages/block-library/src/quote/v2/edit.js:48]: To inform the user that some of the templates they are using need updating? Sounds sensible.

What about ?

```suggestion
		);
		
		deprecated( 'Do not use the value attribute on the quote block', {
			since: '12.10', // To be updated when v2 is the default
			alternative: 'Use inner blocks instead'
		});
```

**Comment by oandregal** [comment]
@youknowriad @ellatrix is there anything else to do here?

File: packages/block-library/src/quote/v2/edit.js (hunk 3 above)
Comment (youknowriad) [reply packages/block-library/src/quote/v2/edit.js:48]:
```suggestion
		);

		deprecated( 'Value attribute on the quote block', {
			since: '6.0', // WP version
			version: '6.5' // when the attribute gets removed. We can drop this property to never remove the deprecation
			alternative: 'inner blocks'
		});
```

File: packages/block-library/src/quote/v2/deprecated.js (hunk 4 above)
Comment (youknowriad) [reply packages/block-library/src/quote/v2/deprecated.js:26]: Replying to myself, this doesn't work for list block because of the nesting.

**Review by youknowriad — CHANGES_REQUESTED** [review]
I've noticed that any format (bold) in paragraphs is removed after conversion.

**Comment by oandregal** [comment]
> I've noticed that any format (bold) in paragraphs is removed after conversion.

The same happens when transforming from pullquote to quote.

File: packages/block-library/src/quote/v2/transforms.js [hunk 6]
```diff
@@ -29,7 +29,7 @@ const transforms = {
 						query: {
 							content: {
 								type: 'string',
-								source: 'text',
```
Comment (oandregal) [inline packages/block-library/src/quote/v2/transforms.js:32]: I've commited this here because it's the same fix I did for the migration. Thought it doesn't deserve a separate PR.

**Comment by oandregal** [comment]
@youknowriad the formats are now working. I've also pushed the same fix for the Pullquote to Quote transform in this PR (https://github.com/WordPress/gutenberg/pull/39844/commits/f3cac0724c2cc8d398efed7fdcb31a97020d7c0b).

**Review by youknowriad — APPROVED** [review]

File: packages/block-library/src/quote/v2/edit.js [hunk 7]
```diff
@@ -9,24 +9,72 @@ import {
...
+
+		deprecated( 'Value attribute on the quote block', {
+			since: '6.0', // WP version
+			version: '6.5' // when the attribute gets removed. We can drop this property to never remove the deprecation
```
Comment (youknowriad) [inline packages/block-library/src/quote/v2/edit.js]:
```suggestion
			version: '6.5'
```

Sorry for that comment, it was mean't to be part of my comment in GitHub :)
